File: backend/app/schemas/approval_flow_config.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


def default_approval_flow_draft() -> dict[str, Any]:
    return {
        "steps": [
            {
                "kind": "approval",
                "title": "环节1",
                "vote_mode": "cosign",
                "assignee_source": "explicit_users",
                "assignee_user_ids": [],
                "dept_id": None,
            }
        ]
    }


class ApprovalFlowVersionCreate(BaseModel):
    label: Optional[str] = Field(None, max_length=200)
    flow: Optional[dict[str, Any]] = Field(default_factory=default_approval_flow_draft)


class ApprovalFlowVersionUpdate(BaseModel):
    label: Optional[str] = Field(None, max_length=200)
    flow: Optional[dict[str, Any]] = None


class ApprovalFlowVersionOut(BaseModel):
    id: int
    project_id: int
    version: int
    label: Optional[str] = None
    status: str
    flow: Optional[dict[str, Any]] = None
    created_by: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/services/approvalFlowConfig.ts
import request from './request'

export type ApprovalFlowVersionRecord = {
  id: number
  project_id: number
  version: number
  label?: string | null
  status: string
  flow: {
    steps: Record<string, unknown>[]
  } | null
  created_by?: number | null
  created_at: string
  updated_at?: string | null
}

export const listApprovalFlowVersions = (projectId: number) =>
  request.get<unknown, ApprovalFlowVersionRecord[]>(
    `/projects/${projectId}/approval-flow-config`,
  )

export const getApprovalFlowVersion = (projectId: number, configId: number) =>
  request.get<unknown, ApprovalFlowVersionRecord>(
    `/projects/${projectId}/approval-flow-config/${configId}`,
  )

export const createApprovalFlowVersion = (
  projectId: number,
  body: { label?: string; flow?: Record<string, unknown> },
) =>
  request.post<unknown, ApprovalFlowVersionRecord>(
    `/projects/${projectId}/approval-flow-config`,
    body,
  )

export const updateApprovalFlowVersion = (
  projectId: number,
  configId: number,
  body: { label?: string; flow?: Record<string, unknown> },
) =>
  request.put<unknown, ApprovalFlowVersionRecord>(
    `/projects/${projectId}/approval-flow-config/${configId}`,
    body,
  )

export const publishApprovalFlowVersion = (projectId: number, configId: number) =>
  request.post<unknown, ApprovalFlowVersionRecord>(
    `/projects/${projectId}/approval-flow-config/${configId}/publish`,
  )
